### PR TITLE
Improve and Update Player Login Process

### PR DESCRIPTION
**Pull request by GoodLucky777**
Improve and Update Player Login Process
- Player
  - StartGamePacket
    - Add a lot of missing data
    - Improve
  - Show player login log after StartGamePacket
  - Add missing packets
  - Add missing attributes in sendAttributes method
  - Add selectedContainerId and SelectedInventorySlot
    - Add related methods: getSelectedContainerId, setSelectedContainerId, getSelectedInventorySlot, setSelectedInventorySlot
  - New method: sendSpawnPosition(int x, int y, int z, int dimension)
- Attribute
  - Update
  - Fix
- New Packet
  - ItemComponentPacket

### CODE REVIEW

File: src/main/java/cn/nukkit/network/protocol/StartGamePacket.java [hunk 1]
```diff
@@ -92,8 +93,8 @@ public byte pid() {
     public boolean commandsEnabled;
     public boolean isTexturePacksRequired = false;
     public GameRules gameRules;
-    public boolean bonusChest = false;
-    public boolean hasStartWithMapEnabled = false;
+    public boolean bonusChestEnabled = false;
+    public boolean startingWithMap = false;
```
Comment (joserobjr) [inline src/main/java/cn/nukkit/network/protocol/StartGamePacket.java:97]: Please, don't rename these fields to preserve binary compatibility with Cloudburst Nukkit

**Review by Polyana-Fontes — DISMISSED** [review]
Thank you for the PR! I just found only one issue, can you adjust it?

File: src/main/java/cn/nukkit/network/protocol/StartGamePacket.java (hunk 1 above)
Comment (good777LUCKY) [reply src/main/java/cn/nukkit/network/protocol/StartGamePacket.java]: ok

**Comment by good777LUCKY** [comment]
Fixed

File: src/main/java/cn/nukkit/Player.java [hunk 2]
```diff
@@ -2179,23 +2179,44 @@ protected void completeLoginSequence() {
...
-        startGamePacket.generator = 1; //0 old, 1 infinite, 2 flat
-        startGamePacket.dimension = (byte) getLevel().getDimension();
-        //startGamePacket.isInventoryServerAuthoritative = true;
+        startGamePacket.worldName = this.getServer().getNetwork().getName(); // Use MOTD
+        startGamePacket.generator = 1; // 0 old, 1 infinite, 2 flat
+        startGamePacket.permissionLevel = PERMISSION_MEMBER;
+        startGamePacket.serverChunkTickRange = 4;
+        startGamePacket.vanillaVersion = "*";
+        startGamePacket.limitedWorldWidth = 16;
+        startGamePacket.limitedWorldHeight = 16;
```
Comment (joserobjr) [inline src/main/java/cn/nukkit/Player.java:2213]: What does this do? Spawn protection?

File: src/main/java/cn/nukkit/Player.java [hunk 3]
```diff
@@ -2179,23 +2179,44 @@ protected void completeLoginSequence() {
...
         startGamePacket.levelId = "";
-        startGamePacket.worldName = this.getServer().getNetwork().getName();
-        startGamePacket.generator = 1; //0 old, 1 infinite, 2 flat
-        startGamePacket.dimension = (byte) getLevel().getDimension();
-        //startGamePacket.isInventoryServerAuthoritative = true;
+        startGamePacket.worldName = this.getServer().getNetwork().getName(); // Use MOTD
+        startGamePacket.generator = 1; // 0 old, 1 infinite, 2 flat
+        startGamePacket.permissionLevel = PERMISSION_MEMBER;
+        startGamePacket.serverChunkTickRange = 4;
+        startGamePacket.vanillaVersion = "*";
```
Comment (joserobjr) [inline src/main/java/cn/nukkit/Player.java:2211]: Why send `*` instead of `ProtocolInfo.MINECRAFT_VERSION_NETWORK`?

**Review by Polyana-Fontes — COMMENTED** [review]
Hi, I have some questions, can you answer? :P

Also, why are you re-assigning the default value to the packet? For example, startGamePacket.multiplayerGame is already true by default, setting it to true again is redundant.

File: src/main/java/cn/nukkit/Player.java (hunk 3 above)
Comment (good777LUCKY) [reply src/main/java/cn/nukkit/Player.java:2211]: I changed to `*` because BDS was using it.

Comment (good777LUCKY) [reply src/main/java/cn/nukkit/Player.java:2211]: I think it is using `baseGameVersion` of level.dat, it was also `*`.

File: src/main/java/cn/nukkit/Player.java (hunk 2 above)
Comment (good777LUCKY) [reply src/main/java/cn/nukkit/Player.java:2213]: I think it is from `limitedWorldWidth`, `limitedWorldDepth` of level.dat. Not sure why is using 16. It was added in Bedrock Edition 1.16 update.

**Comment by good777LUCKY** [comment]
> Hi, I have some questions, can you answer? :P
> 
> Also, why are you re-assigning the default value to the packet? For example, startGamePacket.multiplayerGame is already true by default, setting it to true again is redundant.

Oh, I forgot to remove redundant things. I removed them.

**Comment by GoodLucky777** [comment]
Can someone review this pull request?

**Comment by RORIdev** [comment]
I see what changes joserobjr requested and affirmed that they were all adressed. As such i overwrote his review.

**Comment by GoodLucky777** [comment]
Closing this PR because PN is no longer maintained.

**Pull request closed without merging**